File: exampro/exam_pro/page/exam_dashboard/exam_dashboard.js
frappe.pages['exam-dashboard'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: 'Exam Dashboard',
        single_column: true
    });

    // Add filters
    page.add_field({
        fieldtype: 'Date',
        fieldname: 'start_date',
        label: 'From Date',
        default: frappe.datetime.add_days(frappe.datetime.get_today(), -30),
        onchange: function() {
            exam_dashboard.refresh();
        }
    });
    
    page.add_field({
        fieldtype: 'Date',
        fieldname: 'end_date',
        label: 'To Date',
        default: frappe.datetime.get_today(),
        onchange: function() {
            exam_dashboard.refresh();
        }
    });

    wrapper.exam_dashboard = new ExamDashboard(wrapper, page);
};

// Dashboard class
class ExamDashboard {
    constructor(wrapper, page) {
        this.wrapper = wrapper;
        this.page = page;
        this.make();
    }

    make() {
        this.$container = $('<div class="exam-dashboard">').appendTo(this.wrapper);
        
        // Create sections for metrics, charts, and tables
        this.$metrics_section = $('<div class="metrics-section">').appendTo(this.$container);
        this.$chart_section = $('<div class="chart-section">').appendTo(this.$container);
        this.$table_section = $('<div class="table-section">').appendTo(this.$container);
        
        this.refresh();
    }

    refresh() {
        this.filters = this.get_filters();
        this.fetch_and_render_data();
    }

    get_filters() {
        return {
            start_date: this.page.fields_dict.start_date.get_value(),
            end_date: this.page.fields_dict.end_date.get_value()
        };
    }

    fetch_and_render_data() {
        frappe.call({
            method: "exampro.exam_pro.page.exam_dashboard.exam_dashboard.get_dashboard_data",
            args: {
                filters: this.filters
            },
            callback: (r) => {
                if (!r.exc) {
                    this.render_metrics(r.message);
                    this.render_charts(r.message);
                    this.render_tables(r.message);
                }
            }
        });
    }

    render_metrics(data) {
        const metrics = [
            {
                label: 'Total Exams',
                value: data.total_exams,
                icon: 'book',
                color: 'blue'
            },
            {
                label: 'Total Schedules',
                value: data.total_schedules,
                icon: 'calendar',
                color: 'green'
            },
            {
                label: 'Total Candidates',
                value: data.total_candidates,
                icon: 'users',
                color: 'orange'
            },
            {
                label: 'Completed Exams',
                value: data.completed_exams,
                icon: 'check',
                color: 'purple'
            },
            {
                label: 'Pass Rate',
                value: data.pass_rate + '%',
                icon: 'percentage',
                color: 'red'
            },
            {
                label: 'Avg Score',
                value: data.avg_score.toFixed(2),
                icon: 'star',
                color: 'yellow'
            }
        ];

        this.$metrics_section.empty();
        
        // Create metric cards
        const $metrics_container = $('<div class="metrics-container d-flex flex-wrap">').appendTo(this.$metrics_section);
        
        metrics.forEach(metric => {
            const $metric_card = $(`
                <div class="metric-card" style="width: 200px; margin: 10px; padding: 15px; border-radius: 5px; box-shadow: 0 2px 5px rgba(0,0,0,0.1);">
                    <div class="d-flex align-items-center">
                        <div class="mr-3">
                            <div class="icon-circle bg-${metric.color}-light text-${metric.color}" style="width: 40px; height: 40px; border-radius: 50%; display: flex; align-items: center; justify-content: center;">
                                <i class="fa fa-${metric.icon}"></i>
                            </div>
                        </div>
                        <div>
                            <div class="text-muted small">${metric.label}</div>
                            <div class="metric-value h4 mt-1">${metric.value}</div>
                        </div>
                    </div>
                </div>
            `);
            
            $metrics_container.append($metric_card);
        });
    }

    render_charts(data) {
        this.$chart_section.empty();
        
        // Create chart container with two columns
        const $chart_container = $('<div class="chart-container d-flex flex-wrap">').appendTo(this.$chart_section);
        
        // Create containers for each chart
        const $exam_status_chart = $('<div class="chart-box" style="width: 48%; margin: 1%; height: 300px; box-shadow: 0 2px 5px rgba(0,0,0,0.1); border-radius: 5px; padding: 15px;">').appendTo($chart_container);
        const $exam_submissions_chart = $('<div class="chart-box" style="width: 48%; margin: 1%; height: 300px; box-shadow: 0 2px 5px rgba(0,0,0,0.1); border-radius: 5px; padding: 15px;">').appendTo($chart_container);
        const $candidate_performance_chart = $('<div class="chart-box" style="width: 48%; margin: 1%; height: 300px; box-shadow: 0 2px 5px rgba(0,0,0,0.1); border-radius: 5px; padding: 15px;">').appendTo($chart_container);
        const $schedule_type_chart = $('<div class="chart-box" style="width: 48%; margin: 1%; height: 300px; box-shadow: 0 2px 5px rgba(0,0,0,0.1); border-radius: 5px; padding: 15px;">').appendTo($chart_container);
        
        // Render charts
        this.create_exam_status_chart($exam_status_chart, data);
        this.create_exam_submissions_chart($exam_submissions_chart, data);
        this.create_candidate_performance_chart($candidate_performance_chart, data);
        this.create_schedule_type_chart($schedule_type_chart, data);
    }

    create_exam_status_chart($container, data) {
        $container.append('<h6 class="mb-3">Exam Status Distribution</h6>');
        
        const chart = new frappe.Chart($container[0], {
            data: {
                labels: ["Started", "Not Started", "Submitted"],
                datasets: [{
                    name: "Exam Status",
                    values: [
                        data.status_distribution.Started || 0,
                        data.status_distribution['Not Started'] || 0,
                        data.status_distribution.Submitted || 0
                    ]
                }]
            },
            type: 'pie',
            colors: ['#ff5858', '#5ec962', '#7575ff'],
            height: 240
        });
    }

    create_exam_submissions_chart($container, data) {
        $container.append('<h6 class="mb-3">Exam Submissions Over Time</h6>');
        
        const chart = new frappe.Chart($container[0], {
            data: {
                labels: data.submission_trend.labels,
                datasets: [{
                    name: "Submissions",
                    type: 'line',
                    values: data.submission_trend.values
                }]
            },
            type: 'line',
            colors: ['#7cd6fd'],
            lineOptions: {
                hideDots: 0,
                dotSize: 5
            },
            height: 240
        });
    }

    create_candidate_performance_chart($container, data) {
        $container.append('<h6 class="mb-3">Score Distribution</h6>');
        
        const chart = new frappe.Chart($container[0], {
            data: {
                labels: ["0-20", "21-40", "41-60", "61-80", "81-100"],
                datasets: [{
                    name: "Candidates",
                    values: data.score_distribution
                }]
            },
            type: 'bar',
            colors: ['#5e64ff'],
            height: 240
        });
    }

    create_schedule_type_chart($container, data) {
        $container.append('<h6 class="mb-3">Schedule Types</h6>');
        
        const chart = new frappe.Chart($container[0], {
            data: {
                labels: ["Fixed", "Flexible"],
                datasets: [{
                    name: "Schedule Type",
                    values: [
                        data.schedule_types.Fixed || 0,
                        data.schedule_types.Flexible || 0
                    ]
                }]
            },
            type: 'percentage',
            colors: ['#28a745', '#17a2b8'],
            height: 240
        });
    }

    render_tables(data) {
        this.$table_section.empty();
        
        // Create tables container
        const $tables_container = $('<div class="tables-container">').appendTo(this.$table_section);
        
        // Create Recent Exam Submissions table
        const $recent_submissions = $('<div class="table-box mb-4" style="box-shadow: 0 2px 5px rgba(0,0,0,0.1); border-radius: 5px; padding: 15px;">').appendTo($tables_container);
        $recent_submissions.append('<h6 class="mb-3">Recent Exam Submissions</h6>');
        
        if (data.recent_submissions.length) {
            const $table = $(`
                <table class="table table-bordered">
                    <thead>
                        <tr>
                            <th>Candidate</th>
                            <th>Exam</th>
                            <th>Submitted On</th>
                            <th>Score</th>
                            <th>Status</th>
                        </tr>
                    </thead>
                    <tbody></tbody>
                </table>
            `).appendTo($recent_submissions);
            
            data.recent_submissions.forEach(submission => {
                $(`
                    <tr>
                        <td>${submission.candidate_name}</td>
                        <td>${submission.exam}</td>
                        <td>${submission.submission_time}</td>
                        <td>${submission.score !== null ? submission.score : 'N/A'}</td>
                        <td>
                            <span class="badge ${submission.result_status === 'Pass' ? 'badge-success' : 'badge-danger'}">
                                ${submission.result_status || 'Pending'}
                            </span>
                        </td>
                    </tr>
                `).appendTo($table.find('tbody'));
            });
        } else {
            $recent_submissions.append('<div class="text-muted">No recent submissions</div>');
        }
        
        // Create Top Exams table
        const $top_exams = $('<div class="table-box" style="box-shadow: 0 2px 5px rgba(0,0,0,0.1); border-radius: 5px; padding: 15px;">').appendTo($tables_container);
        $top_exams.append('<h6 class="mb-3">Top Exams by Participation</h6>');
        
        if (data.top_exams.length) {
            const $table = $(`
                <table class="table table-bordered">
                    <thead>
                        <tr>
                            <th>Exam</th>
                            <th>Participants</th>
                            <th>Avg. Score</th>
                            <th>Pass Rate</th>
                        </tr>
                    </thead>
                    <tbody></tbody>
                </table>
            `).appendTo($top_exams);
            
            data.top_exams.forEach(exam => {
                $(`
                    <tr>
                        <td>${exam.exam}</td>
                        <td>${exam.participants}</td>
                        <td>${exam.avg_score ? exam.avg_score.toFixed(2) : 'N/A'}</td>
                        <td>${exam.pass_rate ? exam.pass_rate + '%' : 'N/A'}</td>
                    </tr>
                `).appendTo($table.find('tbody'));
            });
        } else {
            $top_exams.append('<div class="text-muted">No exam data available</div>');
        }
    }
}

File: exampro/exam_pro/page/exam_dashboard/exam_dashboard.py
import frappe
from frappe.utils import getdate, add_days, nowdate, flt
import json

@frappe.whitelist()
def get_dashboard_data(filters=None):
    """
    Get all data for the dashboard
    """
    if isinstance(filters, str):
        filters = json.loads(filters)
    
    if not filters:
        filters = {
            'start_date': add_days(nowdate(), -30),
            'end_date': nowdate()
        }
    
    # Initialize data object
    data = {
        'total_exams': get_total_exams(),
        'total_schedules': get_total_schedules(),
        'total_candidates': get_total_candidates(),
        'completed_exams': get_completed_exams(filters),
        'pass_rate': get_pass_rate(filters),
        'avg_score': get_avg_score(filters),
        'status_distribution': get_status_distribution(filters),
        'submission_trend': get_submission_trend(filters),
        'score_distribution': get_score_distribution(filters),
        'schedule_types': get_schedule_types(filters),
        'recent_submissions': get_recent_submissions(filters),
        'top_exams': get_top_exams(filters)
    }
    
    return data

def get_total_exams():
    """Return total number of exams"""
    return frappe.db.count('Exam')

def get_total_schedules():
    """Return total number of exam schedules"""
    return frappe.db.count('Exam Schedule')

def get_total_candidates():
    """Return total number of candidates"""
    # Count unique candidates in Exam Batch User
    batch_users = frappe.db.sql("""
        SELECT COUNT(DISTINCT candidate) as count 
        FROM `tabExam Batch User`
    """, as_dict=1)
    
    # Count unique candidates in Exam Submission
    submission_users = frappe.db.sql("""
        SELECT COUNT(DISTINCT candidate) as count 
        FROM `tabExam Submission`
    """, as_dict=1)
    
    # Return the higher count as some candidates might be in batches but not have submissions yet
    return max(batch_users[0].count if batch_users else 0, submission_users[0].count if submission_users else 0)

def get_completed_exams(filters):
    """Return number of completed exams within the filter period"""
    return frappe.db.count('Exam Submission', {
        'status': 'Submitted',
        'exam_submitted_time': ['between', [filters.get('start_date'), filters.get('end_date')]]
    })

def get_pass_rate(filters):
    """Calculate pass rate percentage"""
    total_submitted = frappe.db.count('Exam Submission', {
        'status': 'Submitted',
        'exam_submitted_time': ['between', [filters.get('start_date'), filters.get('end_date')]]
    })
    
    if not total_submitted:
        return 0
    
    passed = frappe.db.count('Exam Submission', {
        'status': 'Submitted',
        'result_status': 'Pass',
        'exam_submitted_time': ['between', [filters.get('start_date'), filters.get('end_date')]]
    })
    
    return int(passed / total_submitted * 100) if total_submitted > 0 else 0

def get_avg_score(filters):
    """Calculate average score of submitted exams"""
    avg_score = frappe.db.sql("""
        SELECT AVG(total_marks) as avg_score
        FROM `tabExam Submission`
        WHERE status = 'Submitted'
        AND exam_submitted_time BETWEEN %s AND %s
    """, (filters.get('start_date'), filters.get('end_date')), as_dict=1)
    
    return flt(avg_score[0].avg_score) if avg_score and avg_score[0].avg_score else 0

def get_status_distribution(filters):
    """Get distribution of exam submission statuses"""
    data = frappe.db.sql("""
        SELECT status, COUNT(*) as count
        FROM `tabExam Submission`
        WHERE creation BETWEEN %s AND %s
        GROUP BY status
    """, (filters.get('start_date'), filters.get('end_date')), as_dict=1)
    
    status_dict = {}
    for d in data:
        status_dict[d.status] = d.count
    
    return status_dict

def get_submission_trend(filters):
    """Get exam submission trend over the filter period"""
    start_date = getdate(filters.get('start_date'))
    end_date = getdate(filters.get('end_date'))
    
    # Calculate date intervals based on period length
    date_diff = (end_date - start_date).days
    
    if date_diff <= 7:
        # Daily intervals for periods up to a week
        interval = 1
        format_str = '%d %b'
    elif date_diff <= 31:
        # Every 3 days for a month
        interval = 3
        format_str = '%d %b'
    elif date_diff <= 90:
        # Weekly intervals for periods up to 3 months
        interval = 7
        format_str = '%d %b'
    else:
        # Monthly intervals for longer periods
        interval = 30
        format_str = '%b %Y'
    
    labels = []
    values = []
    
    current_date = start_date
    while current_date <= end_date:
        next_date = add_days(current_date, interval)
        if next_date > end_date:
            next_date = end_date
            
        # Format the label
        labels.append(current_date.strftime(format_str))
        
        # Count submissions in this interval
        count = frappe.db.count('Exam Submission', {
            'exam_submitted_time': ['between', [current_date, next_date]]
        })
        values.append(count)
        
        current_date = add_days(next_date, 1)
    
    return {
        'labels': labels,
        'values': values
    }

def get_score_distribution(filters):
    """Get distribution of scores in ranges"""
    bins = [0, 20, 40, 60, 80, 100]
    distribution = [0, 0, 0, 0, 0]  # For ranges 0-20, 21-40, 41-60, 61-80, 81-100
    
    # Get all scores from submitted exams
    scores = frappe.db.sql("""
        SELECT total_marks
        FROM `tabExam Submission`
        WHERE status = 'Submitted'
        AND exam_submitted_time BETWEEN %s AND %s
    """, (filters.get('start_date'), filters.get('end_date')), as_dict=1)
    
    for score_dict in scores:
        score = score_dict.total_marks
        if score is not None:
            # Place the score in the appropriate bin
            for i in range(len(bins) - 1):
                if bins[i] <= score <= bins[i+1]:
                    distribution[i] += 1
                    break
    
    return distribution

def get_schedule_types(filters):
    """Get distribution of schedule types (Fixed vs Flexible)"""
    data = frappe.db.sql("""
        SELECT schedule_type, COUNT(*) as count
        FROM `tabExam Schedule`
        WHERE creation BETWEEN %s AND %s
        GROUP BY schedule_type
    """, (filters.get('start_date'), filters.get('end_date')), as_dict=1)
    
    type_dict = {}
    for d in data:
        type_dict[d.schedule_type] = d.count
    
    return type_dict

def get_recent_submissions(filters):
    """Get recent exam submissions"""
    submissions = frappe.get_all(
        'Exam Submission',
        filters={
            'status': 'Submitted',
            'exam_submitted_time': ['between', [filters.get('start_date'), filters.get('end_date')]]
        },
        fields=[
            'name', 'candidate', 'candidate_name', 'exam', 
            'exam_submitted_time as submission_time', 
            'total_marks as score', 'result_status'
        ],
        order_by='exam_submitted_time desc',
        limit=10
    )
    
    return submissions

def get_top_exams(filters):
    """Get top exams by participation"""
    # Get exams with submission counts
    exams = frappe.db.sql("""
        SELECT es.exam, COUNT(DISTINCT es.name) as participants,
               AVG(es.total_marks) as avg_score
        FROM `tabExam Submission` es
        WHERE es.creation BETWEEN %s AND %s
        GROUP BY es.exam
        ORDER BY participants DESC
        LIMIT 10
    """, (filters.get('start_date'), filters.get('end_date')), as_dict=1)
    
    # Calculate pass rate for each exam
    for exam in exams:
        total = frappe.db.count('Exam Submission', {
            'exam': exam.exam,
            'status': 'Submitted',
            'creation': ['between', [filters.get('start_date'), filters.get('end_date')]]
        })
        
        if total > 0:
            passed = frappe.db.count('Exam Submission', {
                'exam': exam.exam,
                'status': 'Submitted',
                'result_status': 'Pass',
                'creation': ['between', [filters.get('start_date'), filters.get('end_date')]]
            })
            
            exam['pass_rate'] = int(passed / total * 100)
        else:
            exam['pass_rate'] = 0
    
    return exams
